File: Coocoo3DGraphics/GPUProgram/PSO.cpp
#include "pch.h"
#include "PSO.h"
#include "DirectXHelper.h"
using namespace Coocoo3DGraphics;
static const D3D12_INPUT_ELEMENT_DESC inputLayoutMMD[] =
{
	{ "POSITION", 0, DXGI_FORMAT_R32G32B32_FLOAT, 1, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "NORMAL", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "TEXCOORD", 0, DXGI_FORMAT_R32G32_FLOAT, 0, 12, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "EDGESCALE", 0, DXGI_FORMAT_R32_FLOAT, 0, 20, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "BONES", 0, DXGI_FORMAT_R16G16B16A16_UINT, 0, 24, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "WEIGHTS", 0, DXGI_FORMAT_R32G32B32A32_FLOAT, 0, 32, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "TANGENT", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 48, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
};
static const D3D12_INPUT_ELEMENT_DESC inputLayoutSkinned[] =
{
	{ "POSITION", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "NORMAL", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 12, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "TEXCOORD", 0, DXGI_FORMAT_R32G32_FLOAT, 0, 24, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "TANGENT", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 32, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "EDGESCALE", 0, DXGI_FORMAT_R32_FLOAT, 0, 44, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
};
static const D3D12_INPUT_ELEMENT_DESC inputLayoutPosOnly[] =
{
	{ "POSITION", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
};
static const D3D12_INPUT_ELEMENT_DESC _inputLayoutImGui[] =
{
	{ "POSITION", 0, DXGI_FORMAT_R32G32_FLOAT, 0, 0, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "TEXCOORD", 0, DXGI_FORMAT_R32G32_FLOAT, 0, 8, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
	{ "COLOR", 0, DXGI_FORMAT_R8G8B8A8_UNORM, 0, 16, D3D12_INPUT_CLASSIFICATION_PER_VERTEX_DATA, 0 },
};
inline D3D12_BLEND_DESC BlendDescAlpha()
{
	D3D12_BLEND_DESC blendDescAlpha = {};
	blendDescAlpha.RenderTarget[0].BlendEnable = true;
	blendDescAlpha.RenderTarget[0].SrcBlend = D3D12_BLEND_SRC_ALPHA;
	blendDescAlpha.RenderTarget[0].DestBlend = D3D12_BLEND_INV_SRC_ALPHA;
	blendDescAlpha.RenderTarget[0].BlendOp = D3D12_BLEND_OP_ADD;
	blendDescAlpha.RenderTarget[0].SrcBlendAlpha = D3D12_BLEND_ONE;
	blendDescAlpha.RenderTarget[0].DestBlendAlpha = D3D12_BLEND_INV_SRC_ALPHA;
	blendDescAlpha.RenderTarget[0].BlendOpAlpha = D3D12_BLEND_OP_ADD;
	blendDescAlpha.RenderTarget[0].RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
	return blendDescAlpha;
}
inline D3D12_BLEND_DESC BlendDescAdd()
{
	D3D12_BLEND_DESC blendDescAlpha = {};
	blendDescAlpha.RenderTarget[0].BlendEnable = true;
	blendDescAlpha.RenderTarget[0].SrcBlend = D3D12_BLEND_ONE;
	blendDescAlpha.RenderTarget[0].DestBlend = D3D12_BLEND_ONE;
	blendDescAlpha.RenderTarget[0].BlendOp = D3D12_BLEND_OP_ADD;
	blendDescAlpha.RenderTarget[0].SrcBlendAlpha = D3D12_BLEND_ONE;
	blendDescAlpha.RenderTarget[0].DestBlendAlpha = D3D12_BLEND_INV_SRC_ALPHA;
	blendDescAlpha.RenderTarget[0].BlendOpAlpha = D3D12_BLEND_OP_ADD;
	blendDescAlpha.RenderTarget[0].RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
	return blendDescAlpha;
}
inline D3D12_BLEND_DESC BlendDescSelect(BlendState blendState)
{
	if (blendState == BlendState::none)
		return CD3DX12_BLEND_DESC(D3D12_DEFAULT);
	else if (blendState == BlendState::alpha)
		return BlendDescAlpha();
	else if (blendState == BlendState::add)
		return BlendDescAdd();
	return D3D12_BLEND_DESC{};
}
inline const D3D12_INPUT_LAYOUT_DESC& inputLayoutSelect()
{
	return D3D12_INPUT_LAYOUT_DESC{ inputLayoutSkinned,_countof(inputLayoutSkinned) };
}

void PSO::Initialize(VertexShader^ vs, GeometryShader^ gs, PixelShader^ ps)
{
	ClearState();
	if (vs != nullptr)
		m_vertexShader = vs->byteCode;
	if (gs != nullptr)
		m_geometryShader = gs->byteCode;
	if (ps != nullptr)
		m_pixelShader = ps->byteCode;
	Status = GraphicsObjectStatus::loaded;
}

void PSO::Initialize(const Array<byte>^ vs, const Array<byte>^ gs, const Array<byte>^ ps)
{
	if (vs != nullptr)
	{
		D3DCreateBlob(vs->Length, &m_vertexShader);
		memcpy(m_vertexShader->GetBufferPointer(), vs->begin(), vs->Length);
	}
	if (gs != nullptr)
	{
		D3DCreateBlob(gs->Length, &m_geometryShader);
		memcpy(m_geometryShader->GetBufferPointer(), gs->begin(), gs->Length);

	}
	if (ps != nullptr)
	{
		D3DCreateBlob(ps->Length, &m_pixelShader);
		memcpy(m_pixelShader->GetBufferPointer(), ps->begin(), ps->Length);
	}
}

void PSO::Unload()
{
	ClearState();
	Status = GraphicsObjectStatus::unload;
}

int PSO::GetVariantIndex(GraphicsDevice^ deviceResources, RootSignature^ graphicsSignature, PSODesc psoDesc)
{
	_PSODesc1 _psoDesc1;
	_psoDesc1.desc = psoDesc;
	_psoDesc1.rootSignature = graphicsSignature->m_rootSignature.Get();
	int index = -1;
	for (int i = 0; i < m_psoDescs.size(); i++)
	{
		if (memcmp(&m_psoDescs[i], &_psoDesc1, sizeof(_psoDesc1)) == 0)
		{
			index = i;
		}
	}
	if (index == -1)
	{
		D3D12_SO_DECLARATION_ENTRY declarations[] =
		{
			{0,"POSITION",0,0,3,0},
			{0,"NORMAL",0,0,3,0},
			{0,"TEXCOORD",0,0,2,0},
			{0,"TANGENT",0,0,3,0},
			{0,"EDGESCALE",0,0,1,0},
		};
		UINT bufferStrides[] = { 64 };

		D3D12_GRAPHICS_PIPELINE_STATE_DESC state = {};
		if (psoDesc.inputLayout == InputLayout::mmd)
			state.InputLayout = { inputLayoutMMD, _countof(inputLayoutMMD) };
		else if (psoDesc.inputLayout == InputLayout::postProcess)
			state.InputLayout = { inputLayoutPosOnly, _countof(inputLayoutPosOnly) };
		else if (psoDesc.inputLayout == InputLayout::skinned)
			state.InputLayout = { inputLayoutSkinned, _countof(inputLayoutSkinned) };
		else if (psoDesc.inputLayout == InputLayout::imgui)
			state.InputLayout = { _inputLayoutImGui, _countof(_inputLayoutImGui) };
		state.pRootSignature = graphicsSignature->m_rootSignature.Get();
		if (m_vertexShader != nullptr)
			state.VS = CD3DX12_SHADER_BYTECODE(m_vertexShader.Get());
		if (m_geometryShader != nullptr)
			state.GS = CD3DX12_SHADER_BYTECODE(m_geometryShader.Get());
		if (m_pixelShader != nullptr)
			state.PS = CD3DX12_SHADER_BYTECODE(m_pixelShader.Get());
		if ((DXGI_FORMAT)psoDesc.dsvFormat != DXGI_FORMAT_UNKNOWN)
		{
			state.DepthStencilState = CD3DX12_DEPTH_STENCIL_DESC(D3D12_DEFAULT);
			state.DSVFormat = (DXGI_FORMAT)psoDesc.dsvFormat;
		}
		state.SampleMask = UINT_MAX;
		state.PrimitiveTopologyType = (D3D12_PRIMITIVE_TOPOLOGY_TYPE)psoDesc.ptt;
		if (psoDesc.streamOutput)
		{
			state.StreamOutput.pSODeclaration = declarations;
			state.StreamOutput.NumEntries = _countof(declarations);
			state.StreamOutput.pBufferStrides = bufferStrides;
			state.StreamOutput.NumStrides = _countof(bufferStrides);
		}
		else
		{
			state.BlendState = BlendDescSelect(psoDesc.blendState);
			state.SampleDesc.Count = 1;
		}

		state.NumRenderTargets = psoDesc.renderTargetCount;
		for (int i = 0; i < psoDesc.renderTargetCount; i++)
		{
			state.RTVFormats[i] = (DXGI_FORMAT)psoDesc.rtvFormat;
		}
		D3D12_CULL_MODE cullMode = (D3D12_CULL_MODE)((int)psoDesc.cullMode);
		if (cullMode == 0)cullMode = D3D12_CULL_MODE_NONE;
		state.RasterizerState = CD3DX12_RASTERIZER_DESC(psoDesc.wireFrame ? D3D12_FILL_MODE_WIREFRAME : D3D12_FILL_MODE_SOLID, cullMode, false, psoDesc.depthBias, 0.0f, psoDesc.slopeScaledDepthBias, true, false, false, 0, D3D12_CONSERVATIVE_RASTERIZATION_MODE_OFF);
		Microsoft::WRL::ComPtr<ID3D12PipelineState> pipelineState;
		if (FAILED(deviceResources->GetD3DDevice()->CreateGraphicsPipelineState(&state, IID_PPV_ARGS(&pipelineState))))
		{
			Status = GraphicsObjectStatus::error;
			return -1;
		}
		m_psoDescs.push_back(_psoDesc1);
		m_pipelineStates.push_back(pipelineState);
		return (int)m_psoDescs.size() - 1;
	}
	return index;
}

void PSO::DelayDestroy(GraphicsDevice^ deviceResources)
{
	for (int i = 0; i < m_pipelineStates.size(); i++)
	{
		deviceResources->ResourceDelayRecycle(m_pipelineStates[i]);
	}
}

File: Coocoo3DGraphics1/PSO.cs
using System;
using System.Collections.Generic;
using System.Text;
using Vortice.Direct3D;
using Vortice.Direct3D12;
using Vortice.DXGI;

namespace Coocoo3DGraphics
{
    struct _PSODesc1 : IEquatable<_PSODesc1>
    {
        public PSODesc desc;
        public ID3D12RootSignature rootSignature;

        public override bool Equals(object obj)
        {
            return obj is _PSODesc1 desc && Equals(desc);
        }

        public bool Equals(_PSODesc1 other)
        {
            return desc.Equals(other.desc) &&
                   EqualityComparer<ID3D12RootSignature>.Default.Equals(rootSignature, other.rootSignature);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(desc, rootSignature);
        }

        public static bool operator ==(_PSODesc1 left, _PSODesc1 right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(_PSODesc1 left, _PSODesc1 right)
        {
            return !(left == right);
        }
    }
    public enum BlendState
    {
        none = 0,
        alpha = 1,
        add = 2,
        None = 0,
        Alpha = 1,
        Add = 2,
    };
    public enum InputLayout
    {
        mmd = 0,
        postProcess = 1,
        skinned = 2,
        imgui = 3,
    };

    public struct PSODesc : IEquatable<PSODesc>
    {
        public InputLayout inputLayout;
        public BlendState blendState;
        public CullMode cullMode;
        public PrimitiveTopologyType primitiveTopologyType;
        public Format rtvFormat;
        public Format dsvFormat;
        public int renderTargetCount;
        public int depthBias;
        public float slopeScaledDepthBias;
        public bool wireFrame;

        public override bool Equals(object obj)
        {
            return obj is PSODesc desc && Equals(desc);
        }

        public bool Equals(PSODesc other)
        {
            return inputLayout == other.inputLayout &&
                   blendState == other.blendState &&
                   cullMode == other.cullMode &&
                   primitiveTopologyType == other.primitiveTopologyType &&
                   rtvFormat == other.rtvFormat &&
                   dsvFormat == other.dsvFormat &&
                   renderTargetCount == other.renderTargetCount &&
                   depthBias == other.depthBias &&
                   slopeScaledDepthBias == other.slopeScaledDepthBias &&
                   wireFrame == other.wireFrame;
        }

        public override int GetHashCode()
        {
            HashCode hash = new HashCode();
            hash.Add(inputLayout);
            hash.Add(blendState);
            hash.Add(cullMode);
            hash.Add(primitiveTopologyType);
            hash.Add(rtvFormat);
            hash.Add(dsvFormat);
            hash.Add(renderTargetCount);
            hash.Add(depthBias);
            hash.Add(slopeScaledDepthBias);
            hash.Add(wireFrame);
            return hash.ToHashCode();
        }

        public static bool operator ==(PSODesc left, PSODesc right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(PSODesc left, PSODesc right)
        {
            return !(left == right);
        }
    }
    public class PSO : IDisposable
    {
        static readonly InputLayoutDescription inputLayoutMMD = new InputLayoutDescription(
            new InputElementDescription("POSITION", 0, Format.R32G32B32_Float, 0, 0),
            new InputElementDescription("NORMAL", 0, Format.R32G32B32_Float, 0, 1),
            new InputElementDescription("TEXCOORD", 0, Format.R32G32_Float, 0, 2),
            new InputElementDescription("BONES", 0, Format.R16G16B16A16_UInt, 0, 3),
            new InputElementDescription("WEIGHTS", 0, Format.R32G32B32A32_Float, 0, 4),
            new InputElementDescription("TANGENT", 0, Format.R32G32B32_Float, 0, 5)
            );

        static readonly InputLayoutDescription inputLayoutSkinned = new InputLayoutDescription(
            new InputElementDescription("POSITION", 0, Format.R32G32B32_Float, 0),
            new InputElementDescription("NORMAL", 0, Format.R32G32B32_Float, 0),
            new InputElementDescription("TEXCOORD", 0, Format.R32G32_Float, 0),
            new InputElementDescription("TANGENT", 0, Format.R32G32B32_Float, 0),
            new InputElementDescription("EDGESCALE", 0, Format.R32_Float, 0)
            );
        static readonly InputLayoutDescription inputLayoutPosOnly = new InputLayoutDescription(
            new InputElementDescription("POSITION", 0, Format.R32G32B32_Float, 0)
            );
        static readonly InputLayoutDescription inputLayoutImGui = new InputLayoutDescription(
            new InputElementDescription("POSITION", 0, Format.R32G32_Float, 0),
            new InputElementDescription("TEXCOORD", 0, Format.R32G32_Float, 0),
            new InputElementDescription("COLOR", 0, Format.R8G8B8A8_UNorm, 0)
            );
        static readonly BlendDescription blendStateAdd = new BlendDescription(Blend.One, Blend.One);
        static readonly BlendDescription blendStateAlpha = new BlendDescription(Blend.SourceAlpha, Blend.InverseSourceAlpha, Blend.One, Blend.InverseSourceAlpha);

        BlendDescription BlendDescSelect(BlendState blendState)
        {
            if (blendState == BlendState.none)
                return new BlendDescription(Blend.One, Blend.Zero);
            else if (blendState == BlendState.alpha)
                return blendStateAlpha;
            else if (blendState == BlendState.add)
                return blendStateAdd;
            return new BlendDescription();
        }

        public byte[] vertexShader;
        public byte[] pixelShader;
        public byte[] geometryShader;
        public string Name;
        public GraphicsObjectStatus Status;
        internal List<_PSODesc1> m_psoDescs = new List<_PSODesc1>();
        internal List<ID3D12PipelineState> m_pipelineStates = new List<ID3D12PipelineState>();

        public PSO()
        {

        }

        public PSO(byte[] vertexShader, byte[] geometryShader, byte[] pixelShader)
        {
            this.vertexShader = vertexShader;
            this.geometryShader = geometryShader;
            this.pixelShader = pixelShader;
        }

        public PSO(VertexShader vertexShader, GeometryShader geometryShader, PixelShader pixelShader)
        {
            Initialize(vertexShader, geometryShader, pixelShader);
        }

        public void Initialize(VertexShader vertexShader, GeometryShader geometryShader, PixelShader pixelShader)
        {
            this.vertexShader = vertexShader?.compiledCode;
            this.geometryShader = geometryShader?.compiledCode;
            this.pixelShader = pixelShader?.compiledCode;
        }

        public void Dispose()
        {
            foreach (var combine in m_pipelineStates)
            {
                combine.Release();
            }
            m_psoDescs.Clear();
            m_pipelineStates.Clear();
        }


        internal int GetVariantIndex(GraphicsDevice graphicsDevice, RootSignature graphicsSignature, PSODesc psoDesc)
        {
            _PSODesc1 _psoDesc1;
            _psoDesc1.desc = psoDesc;
            _psoDesc1.rootSignature = graphicsSignature.rootSignature;
            int index = -1;
            for (int i = 0; i < m_psoDescs.Count; i++)
            {
                if (m_psoDescs[i] == _psoDesc1)
                {
                    index = i;
                }
            }
            if (index == -1)
            {
                GraphicsPipelineStateDescription state = new GraphicsPipelineStateDescription();

                //state.InputLayout = inputLayout.inputElementDescriptions;

                if (psoDesc.inputLayout == InputLayout.mmd)
                    state.InputLayout = inputLayoutMMD;
                else if (psoDesc.inputLayout == InputLayout.postProcess)
                    state.InputLayout = inputLayoutPosOnly;
                else if (psoDesc.inputLayout == InputLayout.skinned)
                    state.InputLayout = inputLayoutSkinned;
                else if (psoDesc.inputLayout == InputLayout.imgui)
                    state.InputLayout = inputLayoutImGui;

                state.RootSignature = graphicsSignature.rootSignature;
                if (vertexShader != null)
                    state.VertexShader = vertexShader;
                if (geometryShader != null)
                    state.GeometryShader = geometryShader;
                if (pixelShader != null)
                    state.PixelShader = pixelShader;
                state.SampleMask = uint.MaxValue;
                state.PrimitiveTopologyType = psoDesc.primitiveTopologyType;
                state.BlendState = BlendDescSelect(psoDesc.blendState);
                state.SampleDescription = new SampleDescription(1, 0);

                state.RenderTargetFormats = new Format[psoDesc.renderTargetCount];
                for (int i = 0; i < psoDesc.renderTargetCount; i++)
                {
                    state.RenderTargetFormats[i] = psoDesc.rtvFormat;
                }
                CullMode cullMode = psoDesc.cullMode;
                if (cullMode == 0) cullMode = CullMode.None;
                RasterizerDescription rasterizerDescription = new RasterizerDescription(cullMode, psoDesc.wireFrame ? FillMode.Wireframe : FillMode.Solid);
                rasterizerDescription.DepthBias = psoDesc.depthBias;
                rasterizerDescription.SlopeScaledDepthBias = psoDesc.slopeScaledDepthBias;
                if (psoDesc.dsvFormat != Format.Unknown)
                {
                    state.DepthStencilState = new DepthStencilDescription(true, true);
                    state.DepthStencilFormat = psoDesc.dsvFormat;
                    rasterizerDescription.DepthClipEnable = true;
                }
                else
                {
                    state.DepthStencilState = new DepthStencilDescription(false, false);
                }

                state.RasterizerState = rasterizerDescription;
                ID3D12PipelineState pipelineState;
                if (graphicsDevice.device.CreateGraphicsPipelineState(state, out pipelineState).Failure)
                {
                    Status = GraphicsObjectStatus.error;
                    return -1;
                }
                m_psoDescs.Add(_psoDesc1);
                m_pipelineStates.Add(pipelineState);
                return (int)m_psoDescs.Count - 1;
            }
            return index;
        }
    }
}
